src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -137,23 +137,34 @@ public override BoundNode VisitCall(BoundCall node)
                     bool mightMutate =
                         // might be a call to a local function that assigns something
                         node.Method.MethodKind == MethodKind.LocalFunction ||
-                        // or perhaps we are passing a variable by ref and mutating it that way
+                        // or perhaps we are passing a variable by ref and mutating it that way, e.g. `int.Parse(..., out x)`
                         !node.ArgumentRefKindsOpt.IsDefault ||
                         // or perhaps we are calling a mutating method of a value type
-                        !node.Method.IsStatic && node.Method.ContainingType.IsStructType() && node.Method.ContainingType.SpecialType == SpecialType.None;
+                        MethodMayMutateRecevier(node.Method);
 
                     if (mightMutate)
                         _mightAssignSomething = true;
                     else
                         base.VisitCall(node);
-
                     return null;
                 }
 
+                private static bool MethodMayMutateRecevier(MethodSymbol method)
+                {
+                    return
+                        method != null &&
+                        !method.IsStatic &&
+                        method.ContainingType.IsStructType() &&
+                        // methods of special types do not mutate their receiver
+                        method.ContainingType.SpecialType == SpecialType.None &&
+                        !method.IsDeclaredReadOnly;
```

Comment:
`!method.IsEffectivelyReadOnly`? e.g. if the method were contained in a `readonly struct` declaration.

Reply:
(I should probably revisit whether `IsDeclaredReadOnly` should be made `protected` and rename `IsEffectivelyReadOnly` to just `IsReadOnly` in the internal API.)